gr/include/gnuradio/block_work_io.h

```diff
@@ -130,4 +143,44 @@ enum class work_return_code_t {
            /// parent interface when it is ready to be called again
 };
 
+
+class work_io
+{
+public:
+    friend class block;
+    size_t nin() { return _inputs.size(); }
+    size_t nout() { return _outputs.size(); }
+    block_work_input_sptr& input(size_t idx)
+    {
+        return _inputs[idx];
+    }
+    block_work_input_sptr& input(const std::string&);
+    block_work_output_sptr& output(size_t idx)
+    {
+        return _outputs[idx];
+    }
+    block_work_output_sptr& output(const std::string&);
+    std::vector<block_work_input_sptr>& inputs() { return _inputs; }
+    std::vector<block_work_output_sptr>& outputs() { return _outputs; }
+    // block_work_input_sptr& operator[](size_t);
+    // block_work_input_sptr& operator[](const std::string&);
+
+    void consume_each(size_t n_items) {
+        for (auto& w : inputs()) {
+            w->n_consumed = n_items;
+        }
+    }
+    void produce_each(size_t n_items) {
+        for (auto& w : outputs()) {
+            w->n_produced = n_items;
+        }
+    }
+private:
+    std::vector<block_work_input_sptr> _inputs;
+    std::vector<block_work_output_sptr> _outputs;
+
+    std::map<std::string, block_work_input_sptr> _input_name_map;
```

Comment:
Definitely. This map should be from string to object. The object will not be destroyed/recreated during the lifetime of this mapping (that is, the `shared_ptr` won't start pointing to something else), so pointing directly to the object seems appropriate.

And I think an important thing to try to do in newsched is to reduce the use of `shared_ptr` (though I've not kept up much with its development lately).

If `work_io` needs to be movable then raw objects in the vector requires those blocks to be movable, and requires `work_io` to implement a custom move construct/assign. So may be best to have `unique_ptr` in the vector.

Either way using a `shared_ptr` tells the reader that the design doesn't know who owns the objects, which is not the case.

This applies to the return value of `input()` and `output()` too. Why are they smart pointers? Why are they even pointers? Should should surely be references to the objects? `work_io` is not sharing ownership of the object with the caller, is it?

Reply:
Let me clean up the shared ptrs in this context.  I had used them when the vectors of shared_ptr were being passed directly into the work method, which was also not necessary per your comments, but now as an object in a higher level object, even less so.

Reply:
I have a tendency to abuse shared_ptrs :/

Reply:
It's not straight up abuse – starting with the most universal pointer and then refining one's design, that's the way of the craft. Sometimes ownerships are clear up front, sometimes you need to write code to see what lives where :)

Reply:
In my opinion that's the wrong place to start, and makes it easy to paint yourself into a corner.

Start with not a pointer at all, then `unique_ptr`.

Otherwise it's easy to run very far down the corridor and write lots of code, only to see that no this won't work at all and have to undo everything.

Reply:
hm, how does that work? If your vector is holding the object, and your map is holding the unique_ptr, then destruction of both containers at end of instance life time will double-deconstruct the object.

Reply:
oof .. yeah have not fully thought this through yet apparently.  how would you do it?

Reply:
If you should stick with both the vector and the map (sounds like you're trying without the map, and if it becomes a problem we can add the map back on a  case-by-case basis if  the vector gets too big in a future patch), hold the actual objects in one container, and non-smart pointers to them in the other. Destructing dumb pointers doesn't trigger destruction!